src/uct/sm/mm/mm_ep.c

```diff
@@ -58,10 +84,72 @@ ucs_status_t uct_mm_ep_put_bcopy(uct_ep_h tl_ep, uct_pack_callback_t pack_cb,
     return UCS_OK;
 }
 
-ucs_status_t uct_mm_ep_am_short(uct_ep_h ep, uint8_t id, uint64_t header,
+static inline ucs_status_t uct_mm_ep_get_remote_elem(uct_mm_ep_t *ep, uint64_t head,
+		                                             uct_mm_fifo_element_t **elem)
+{
+    uct_mm_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_mm_iface_t);
+    uint64_t elem_index;       /* the fifo elem's index in the fifo. */
+                               /* must be smaller than fifo size */
+    uint64_t returned_val;
+
+    elem_index = ep->fifo_ctl->head & iface->fifo_mask;
+    (*elem) = UCT_MM_IFACE_GET_FIFO_ELEM(iface, ep->fifo, elem_index);
+
+    /* try to get ownership of the head element */
+    returned_val = ucs_atomic_cswap64(&ep->fifo_ctl->head, head, head+1);
```

Comment:
ucs_atomic_cswap64 gcc implementation on POWER does more than needed for updating the ep->fifo_ctl->head. gcc takes the too safe approach with isync + instructions + lwsync. In this case, only a lwsync is needed.